src/Adapters/Auth/AuthAdapter.js

```diff
@@ -1,20 +1,102 @@
 /*eslint no-unused-vars: "off"*/
 export class AuthAdapter {
-  /*
+  constructor() {
+    /**
+     * Usage policy
+     * default: can be combined with ONE additional auth provider if additional configured on user
+     * additional: could be only used with a default policy auth provider
+     * solo: Will ignore ALL additional providers if additional configured on user
+     */
+    this.policy = 'default';
+  }
+  /**
   @param appIds: the specified app ids in the configuration
   @param authData: the client provided authData
   @param options: additional options
+  @param req: RestWrite instance with config/auth/data
   @returns a promise that resolves if the applicationId is valid
    */
-  validateAppId(appIds, authData, options) {
+  validateAppId(appIds, authData, options, req) {
+    return Promise.resolve({});
+  }
+
+  /**
+   * Legacy usage, if provided it will be triggered when authData related to this provider is touched (signup/update/login)
+   * otherwise you should implement validateSetup, validateLogin and validateUpdate
+  @param authData: the client provided authData
+  @param options: additional options
+  @param req: RestWrite instance with config/auth/data
+  @param user: Parse.User instance if Parse.User found
+  @returns a promise that resolves, the resolved value will be handled by the server like:
+  - resolve undefined|void|{} parse server will save authData
+  - resolve { doNotSave: boolean, response: Object} parse server will do not save provided authData and send response to the client under authDataResponse
```

Comment:
Prefer

`- resolve { doNotSave: boolean, response: Object} Provided authData will not be saved, and response will be sent to the client under authDataResponse`

Reply:
I'm not sure but maybe we should always write code in single backticks (`code`), because that may also format properly as code in IDEs and doc compilers.

Reply:
JSDoc issues addressed in separate review comments